via-router/src/binding.rs

```diff
@@ -47,13 +43,36 @@ impl<T> Binding<'_, T> {
 
 impl<'a, T> Binding<'a, T> {
     #[inline]
-    pub(crate) fn new(range: Option<[usize; 2]>, nodes: SmallVec<[MatchKind<'a, T>; 1]>) -> Self {
-        Self { range, nodes }
+    pub(crate) fn new(range: [usize; 2]) -> Self {
+        Self {
+            has_nodes: false,
+            nodes: SmallVec::new(),
+            range: Some(range),
+        }
+    }
+
+    pub(crate) fn new_with_nodes(
+        range: Option<[usize; 2]>,
+        nodes: SmallVec<[MatchKind<'a, T>; 1]>,
+    ) -> Self {
+        debug_assert!(
+            !nodes.is_empty(),
+            "Binding::new_with_nodes requires that nodes is not empty"
+        );
+
+        Self {
+            has_nodes: true,
+            nodes,
+            range,
+        }
     }
 
     #[inline]
     pub(crate) fn push(&mut self, node: MatchKind<'a, T>) {
         self.nodes.push(node);
+        if !self.has_nodes {
```

Comment:
I was debating whether or not I should keep the branch here. I'm leaning towards unconditionally setting has_nodes to decrease the number of instructions